onnx/defs/nn/defs.cc

```diff
@@ -1718,10 +1750,58 @@ ONNX_OPERATOR_SET_SCHEMA(
             "T",
             {"tensor(float16)", "tensor(float)", "tensor(double)"},
             "Constrain input and output types to float tensors.")
+        .TypeConstraint(
+            "T1",
+            {"tensor(bool)"},
+            "Constrain input training_mode to boolean tensors.")
         .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
           propagateShapeAndTypeFromFirstInput(ctx);
-          // TODO in training mode, it may be possible to infer some of
-          // the other outputs as well.
+          propagateShapeFromInputToOutput(ctx, 0, 0);
+
+          Dim num_channels;
+
+          unifyInputDim(ctx, 0, 1, num_channels);
+          unifyInputDim(ctx, 1, 0, num_channels);
+          unifyInputDim(ctx, 2, 0, num_channels);
+          unifyInputDim(ctx, 3, 0, num_channels);
+          unifyInputDim(ctx, 4, 0, num_channels);
+
+          if (ctx.getNumInputs() > 5 && hasInputShape(ctx, 5)) {
+            auto& mode_input_shape = getInputShape(ctx, 5);
+            // if mode is not scalar or tensor of rank 1, fail shape inference
```

Comment:
I think that for new op extensions, it is preferable to restrict scalars to be of rank 0.

@postrational what do you think? This is one of the "unified op interface" items. Traditionally, there has been some divergence with some ops using or allowing rank-1 tensors of size 1 for a scalar, with some requiring rank-0 tensors.